File: leetcode/py/p66-plus-one.py
"""LeetCode 66. Plus One.
"""

from typing import List
import copy

class Solution:
    """dumb docstring"""
    def plus_one(self, digits : List[int]) -> List[int]:
        for x in range(len(digits) - 1, -1, -1):
            digits[x] += 1
            if digits[x] < 10:
                return digits
            else:
                digits[x] = 0
        digits.insert(0, 1)
        return digits

test_cases = (([0], [1]),
    ([9], [1,0]),
    ([8,9], [9,0]),
    ([9,9], [1, 0,0]),
    ([1,2,3], [1,2,4]),
    ([4,3,2,1], [4,3,2,2]),
)

solution = Solution()
for digits_, exp in test_cases:
    digits_copy = copy.deepcopy(digits_)
    act = solution.plus_one(digits_copy)
    if exp != act:
        print(f"digits: {digits_copy}, exp: {exp}, act: {act}")

File: leetcode/rust/p66-plus-one.rs
struct Solution {}

impl Solution {
    pub fn plus_one(digits: Vec<i32>) -> Vec<i32> {
        let mut digits = digits;
        for i in (0..digits.len()).rev() {
            digits[i] += 1;
            if digits[i] < 10 {
                return digits;
            }
            digits[i] = 0;
        }
        digits.insert(0, 1);
        return digits;
    }
}

fn main() {
    struct TestCase {
        digits: Vec<i32>,
        exp: Vec<i32>,
    }

    let test_cases = vec![
        TestCase {
            digits: vec![0],
            exp: vec![1],
        },
        TestCase {
            digits: vec![9],
            exp: vec![1, 0],
        },
        TestCase {
            digits: vec![8, 9],
            exp: vec![9, 0],
        },
        TestCase {
            digits: vec![9, 8],
            exp: vec![9, 9],
        },
        TestCase {
            digits: vec![9, 9],
            exp: vec![1, 0, 0],
        },
        TestCase {
            digits: vec![9, 9],
            exp: vec![1, 0, 0],
        },
    ];

    for tc in test_cases.iter() {
        let act = Solution::plus_one(tc.digits.clone());
        if tc.exp != act {
            println!(
                "Fail, digits: {:?}, exp: {:?}, act: {:?}",
                tc.digits, tc.exp, act
            )
        }
    }
}
